### PR TITLE
PLT-250: mkEvaluationContext now lives in (MonadError CostModelApplyError)

### PR DESCRIPTION
**Pull request by bezirg**
Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs [hunk 1]
```diff
@@ -88,9 +90,24 @@ extractParams cm = case toJSON cm of
     _ -> Nothing
 
 
+data CostModelApplyError =
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs:93]: Haddock

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs [hunk 2]
```diff
@@ -88,9 +90,24 @@ extractParams cm = case toJSON cm of
...
+      CMUnknownParamError Text.Text
+    | CMReadInputError
+    | CMGenOutputError
+instance Exception CostModelApplyError
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs:97]: do we want this?

Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs:97]: I am using it somewhere for convenience , so why not.

File: plutus-ledger-api/src/Plutus/ApiCommon.hs [hunk 3]
```diff
@@ -268,15 +270,15 @@ data EvaluationContext = EvaluationContext
...
 -}
-mkEvaluationContext :: Plutus.CostModelParams -> Maybe EvaluationContext
+mkEvaluationContext :: MonadError CostModelApplyError m => Plutus.CostModelParams -> m EvaluationContext
 mkEvaluationContext newCMP =
     EvaluationContext
         <$> inline mkMachineParametersFor UnliftingImmediate newCMP
         <*> inline mkMachineParametersFor UnliftingDeferred newCMP
 
 -- | Comparably expensive to `mkEvaluationContext`, so it should only be used sparingly.
 isCostModelParamsWellFormed :: Plutus.CostModelParams -> Bool
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-ledger-api/src/Plutus/ApiCommon.hs:280]: u mean i should instantiate the monaderror?

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs [hunk 4]
```diff
@@ -102,14 +119,14 @@ applyParams cm params = case toJSON cm of
             merged <- Map.mergeA failMissing Map.preserveMissing (Map.zipWithMatched leftBiased) usingScientific flattened
             let unflattened = unflattenObject "-" $ hmToObj $ toHash merged
             case fromJSON (Object unflattened) of
-                Success a -> Just a
-                Error _   -> Nothing
-    _ -> Nothing
+                Success a -> pure a
+                Error _   -> throwError CMGenOutputError
+    _ -> throwError CMReadInputError
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs:134]: yeah, I mean maybe just the error can be more descriptive

File: plutus-ledger-api/src/Plutus/ApiCommon.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-ledger-api/src/Plutus/ApiCommon.hs:280]: oh, running in `MonadError` is also fine, I just mean it's nice to get more than the bool. Although `MonadError m => ... -> m Bool` is a weird type :D

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs [hunk 5]
```diff
@@ -88,9 +92,31 @@ extractParams cm = case toJSON cm of
...
+      -- ^ internal error when we are transforming the applyParams' input to json (should not happen)
+    | CMInternalWriteError String
+      -- ^ internal error when we are transforming the applied params from json with given jsonstring error (should not happen)
+    deriving stock Show
+    deriving anyclass Exception
+
+instance Pretty CostModelApplyError where
+    pretty = (preamble <+>) . \case
+        CMUnknownParamError k -> "Unknown cost model parameter:" <+> pretty k
+        CMInternalReadError      -> "internal problem occurred upon reading the given costmodel params"
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs:109]:
```suggestion
        CMInternalReadError      -> "Internal problem occurred upon reading the given cost model parameters"
```

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs [hunk 6]
```diff
@@ -88,9 +92,31 @@ extractParams cm = case toJSON cm of
...
+        CMInternalWriteError str     -> "internal problem occurred upon generating the applied costmodel params with JSON error:" <+> pretty str
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs:110]:
```suggestion
        CMInternalWriteError str     -> "Internal problem occurred upon generating the applied cost model parameters, with JSON error:" <+> pretty str
```

**Review by michaelpj — APPROVED** [review]
Just a couple of wording comments.

**Comment by michaelpj** [comment]
Also genuine compiler error apparently.

File: plutus-ledger-api/src/Plutus/V1/Ledger/Api.hs [hunk 7]
```diff
@@ -24,8 +24,9 @@ module Plutus.V1.Ledger.Api (
     -- ** Cost model
     , EvaluationContext
     , mkEvaluationContext
+    , CostModelApplyError
```
Comment (michaelpj) [inline plutus-ledger-api/src/Plutus/V1/Ledger/Api.hs:27]: and constructors!

**Pull request merged**